kythe/go/extractors/govname/govname.go

```diff
@@ -152,3 +151,57 @@ func ForStandardLibrary(importPath string) *spb.VName {
 func IsStandardLibrary(v *spb.VName) bool {
 	return v != nil && (v.Language == "go" || v.Language == "") && v.Corpus == golangCorpus
 }
+
+// RepoRoot analyzes importPath to determine it's vcs.RepoRoot.
+func RepoRoot(importPath string) (*vcs.RepoRoot, error) {
+	if root := repoRootCache.lookup(strings.Split(importPath, "/")); root != nil {
+		return root, nil
+	}
+	r, err := vcs.RepoRootForImportPath(importPath, false)
+	if err != nil {
+		return nil, err
+	}
+	repoRootCache.add(strings.Split(r.Root, "/"), r)
+	return r, nil
+}
+
+var repoRootCache repoRootCacheNode
+
+// repoRootCacheNode is a prefix search tree node for *vcs.RepoRoots by their
+// root import path.
+type repoRootCacheNode struct {
+	root *vcs.RepoRoot
+
+	children map[string]*repoRootCacheNode
+}
+
+// add puts the given *vcs.RepoRoot into the prefix tree for the given path
+// components.
+func (n *repoRootCacheNode) add(components []string, r *vcs.RepoRoot) {
+	if len(components) == 0 {
```

Comment:
so to be clear, a second call to .add() with another empty components would reset root, and that's expected?

Reply:
Yup.